### ISSUE TITLE
Deploy built app to Azure from Travis

### ISSUE THREAD

**Issue by martintoreilly** [alan-turing-institute/science-gateway-inputs#12]
## Description
The front-end app has a build stage so we want to deploy the built artifacts to Azure and would prefer not to check these into our source repository. Options are:

- Separate builds on Travis CI and Azure.
  - Relies on protected nature of `master` brach so that Travis CI passes prior to merge into `master`. We then have Azure Kudu pull from our source repo and build the app itself.
  - Challenge is to get reasonable assurance that we are building exactly the same thing on Travis and Azure.
- Push the built app from Travis to Azure
  - Guarantees that the built app has passed CI
  - Need to ensure we don't run into any issues with incompatible versions of things on Travis and Azure.

## Implementation
We have decided to deploy from Travis. It should be as simple as adding the following to `.travis.yml` with the following environmental variables set in Travis. This should get Travis to push the repo including the build output to the Azure local Git repo for the app service.

### Travis Yaml
```yml
before_deploy: rm .gitignore
deploy:
  provider: azure_web_apps
  verbose: true
  skip_cleanup: true
  on: 
    all_branches: true 
after_deploy: echo 'done deploying!'
```

Removing `.gitignore` before deploy and skipping cleanup ensures that the `dist` build output directory is pushed to Azure.

Deploy set to occur on all branches for testing but should be `master` only once successful. We should also look at having deployment slots for dev / staging / production, linked to different source branches in Travis.

### Travis environmental variables
`AZURE_WA_SITE` set to subdomain of app on `azurewebistes.net` (i.e. `science-gateway-inputs` for us)
`AZURE_WA_USERNAME` set to Azure app service deployment username (used for FTP and local Git deployment)
`AZURE_WA_PASSWORD` set to Azure app service deployment password (used for FTP and local Git deployment)

**Comment by martintoreilly** [comment]
We are getting an authentication error on deployment:
`fatal: Authentication failed for 'https://hut23-deploy:[secure]@science-gateway-inputs.scm.azurewebsites.net:443/science-gateway-inputs.git/'`

We have verified that the username and password are the same on the Azure app service and Travis.

We could look at adding verbose logging to the deployment by adding `verbose: true` to the `deploy:` section in `travis.yml` ([source](https://docs.travis-ci.com/user/deployment/azure-web-apps/#Fetch-Deployment-Progress-and-Logs))

**Comment by dpshelio** [comment]
Problem solved!! passwords with special characters need to be escaped on Travis. Messaged shown in Travis:

> Notice that the values are not escaped when your builds are executed. Special characters (for bash) should be escaped accordingly.

**Comment by dpshelio** [comment]
Ha!! and `$` [needs to be escaped with 3`\`: `\\\`](https://stackoverflow.com/a/42608990/1087595) - Now it works!!
